src/Microsoft.VisualStudio.Composition/Configuration/Formatters/MessagePackSerializerContext.cs

```diff
@@ -104,30 +104,25 @@ internal MyDedupingResolver(IFormatterResolver inner)
         private class DedupingFormatter<T> : IMessagePackFormatter<T>
         {
             private readonly MyDedupingResolver owner;
+            private readonly bool isValueType;
 
             internal DedupingFormatter(MyDedupingResolver owner)
             {
                 this.owner = owner;
+                this.isValueType = typeof(T).IsValueType;
             }
 
             public T Deserialize(ref MessagePackReader reader, MessagePackSerializerOptions options)
             {
-                if (!typeof(T).IsValueType && reader.TryReadNil())
+                if (!this.isValueType && reader.TryReadNil())
                 {
                     return default!;
                 }
 
                 if (reader.NextMessagePackType == MessagePackType.Extension)
                 {
-                    MessagePackReader provisionaryReader = reader.CreatePeekReader();
-                    ExtensionHeader extensionHeader = provisionaryReader.ReadExtensionFormatHeader();
-                    if (extensionHeader.TypeCode == ReferenceExtensionTypeCode)
-                    {
-                        int id = provisionaryReader.ReadInt32();
-                        reader = provisionaryReader;
-
-                        return (T)(this.owner.deserializedObjects[id] ?? throw new MessagePackSerializationException("Unexpected null element in shared object array. Dependency cycle?"));
-                    }
+                    reader.ReadExtensionFormatHeader();
```

Comment:
Your change here discards the extension typecode and reads 4 bytes, assuming it matches. This will malfunction when the value to be deserialized is a different extension than the one used by this particular formatter (which is allowed).